feat: Add movie recommendation system using Slope One algorithm

VaDiM-2012 opened the pull request.

VaDiM-2012 committed:
feat: Add movie recommendation system using Slope One algorithm

Implemented a recommendation system for films.
- Added RecommendationService with methods to compute film scores and generate top-10 recommendations.
- Updated UserController with GET /users/{id}/recommendations endpoint.
- Extended FilmLikesRepository and FilmLikesDbRepository with getAllLikes method.
- Introduced FilmLike model to represent user-film like pairs.
- Added unit tests in RecommendationServiceTest using 7 users to verify recommendation logic.

worder reviewed (changes requested):
Решение рабочее, но мне почему-то кажется ревьювер не примет такую реализацию. Тут большие проблемы с масштабируемостью, так как почти вся база выгружается в память для вычислений. Вся эта логика заменяется одним sql запросом и проблема с масштабируемостью пропадает, так как БД делает все эти операции на ограниченных выборках. Попробуешь сделать на бд? Времени еще много до дедлайнов. Если что, можно обсудить здесь или в тг.

VaDiM-2012 committed:
Refactored RecommendationService to use SQL query for computing film recommendations:
- Replaced in-memory processing with a single SQL query to calculate film scores based on user like intersections.
- Updated unit tests in RecommendationServiceTest to validate SQL-based implementation with 7 users.

worder commented on src/main/java/ru/yandex/practicum/filmorate/service/RecommendationService.java:
```diff
@@ -32,37 +32,32 @@ public List<FilmDto> getRecommendations(Long userId) {
+                  AND fl2.film_id NOT IN (SELECT film_id FROM film_likes WHERE user_id = ?)
+                GROUP BY fl2.film_id
+                ORDER BY score DESC
+                LIMIT 10;
```
а разве было требование такое?

worder commented on src/main/java/ru/yandex/practicum/filmorate/service/RecommendationService.java:
```diff
@@ -32,37 +32,32 @@ public List<FilmDto> getRecommendations(Long userId) {
 
         // Преобразуем ID фильмов в DTO
-        List<FilmDto> recommendations = getFilmDtos(recommendedFilmIds);
+        List<FilmDto> recommendations = recommendedFilmIds.stream().map(filmService::getFilmById).toList();
```
если лимит убрать, здесь будут запросы в цикле, ревьюверы такое не любят

worder replied:
Вообще, как мне кажется, сервис теперь излишний для одного запроса, тем более мы запросы в сервисах не делаем. Думаю, стоит переместить этот функционал в filmDbStorage, и сделать метод в filmservice

worder replied:
как для популярных фильмов

VaDiM-2012 committed:
- The RecommendationService service has been removed.
- The logic of getting recommendations has been moved to FilmDbRepository, similar to the implementation of popular films.
- Optimized SQL query has been added

VaDiM-2012 committed:
Merge remote-tracking branch 'origin/develop' into add-recommendations

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java
#	src/main/java/ru/yandex/practicum/filmorate/dal/FilmRepository.java
#	src/main/java/ru/yandex/practicum/filmorate/dal/inmemory/InMemoryFilmRepository.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java

worder commented on src/main/java/ru/yandex/practicum/filmorate/dal/db/FilmDbRepository.java:
```diff
@@ -45,6 +45,19 @@ INSERT INTO films (name, description, release_date, duration, mpa_rating_id)
+            GROUP BY fl2.film_id
+            ORDER BY score DESC
+            """;
+
```
вот теперь хорошо)

worder reviewed (approved):
принял
